### PR TITLE
Add semaphore pool

### PR DESCRIPTION
**Pull request by CodingRays**
This is just a small utility class providing a semaphore pool. The idea is that any time some operation requires a semaphore it can get one from the pool and then return it after the operation has completed. (for example using a fence and asynchronous callback or some other way). The benefit is that the user doesn't have to worry about ensuring that the operation has completed before resubmitting the semaphore again which can be a headache when you use locally owned semaphores.

### CODE REVIEW

**Comment by OroArmor** [comment]
what is the use case for this?

**Comment by hYdos** [comment]
> what is the use case for this?

> 
 The benefit is that the user doesn't have to worry about ensuring that the operation has completed before resubmitting the semaphore again which can be a headache when you use locally owned semaphores.

**Comment by CodingRays** [comment]
2 main use cases:
1. You have a chain of submissions that are dependant on each other (happens often when you are transferring data). So you have to ensure ordering using semaphores. Then being able to just quickly allocate as many semaphores as you need can be a lot easier and faster than submitting one command after the other and waiting for it to finish so that you can submit the next safely. 
2. Even for recurring dependencies each frame this can be easier since once you go full async it is possible that part of the pipeline is already working on the next frame while the last frame is still being rendered. In this case ensuring that the semaphore from the last frame is safe to use can be annoying so just requesting a new one from the pool is much simpler and less error prone. All of the semaphores from the last frame can then be returned to the pool with a simple fence + async waiting thread at the end of the frame.

File: Rosella/src/main/java/me/hydos/rosella/util/SemaphorePool.java
Comment (hYdos) [inline Rosella/src/main/java/me/hydos/rosella/util/SemaphorePool.java:17]: is there a reason behind this number? just curious. its not a problem

**Review by hYdos — APPROVED** [review]

**Comment by hYdos** [comment]
all good code wise. will wait for @OroArmor to approve before merging

File: Rosella/src/main/java/me/hydos/rosella/util/SemaphorePool.java
Comment (CodingRays) [reply Rosella/src/main/java/me/hydos/rosella/util/SemaphorePool.java:17]: No not really i just had to pick some random number and 16 looked nice.

**Comment by CodingRays** [comment]
Well the checks are failing though i have no clue as to why

**Comment by CodingRays** [comment]
Can the checks be rerun? They seemed to fail because some online service wasnt available (or at least thats what i thing looking at the error `ECONNRESET`). I havent changed anything in gradle so i dont see anything else that could be wrong...

**Pull request merged**
